rfd/0082-session-tracker-resource-rbac.md

```diff
@@ -0,0 +1,69 @@
+---
+authors: Joel Wejdenstål (jwejdenstal@goteleport.com)
+state: draft
+---
+
+# RFD 82 - Session Tracker Resource RBAC
+
+## What
+
+Introduce a new resource kind `session_tracker` that may be used in the RBAC verb system
+to manage access for existing session tracker resources using list/read resource verbs in addition
+to dynamic `where` conditions in the same manner as RFD 44.
+
+## Why
+
+Session tracker visibility is currently tightly coupled with the permission to join a given session.
+They are invisible by default and may only be read by users with permission to join a session in some mode
+using the `join_sessions` primitive from Moderated Sessions. The functionality proposed by this RFD will allow
+one to configure access to list/read session tracker resources without being able to join the sessions in question.
+
+## Details
+
+### API
+
+Only the `list` and `read` verbs are supported for session tracker resources. Supporting other verbs makes little
+sense as tracker resources may only be safely managed by the node that created them. The `list` verb will allow access
+to the `GetActiveSessionTrackers` RPC while `read` will provide access to the `GetSessionTracker` RPC.
+
+The `where` clause may be used to specify an expression that will be used to selectively filter or allow/deny a request
+based on tracker contents.
+
+The environment in which the expression will execute defines the following variables derived from the tracker object and request user:
+- `user`: standard user object as defined in RFD 44
+- `tracker`: object
+  - `session_id`: string
+  - `kind`: string
+  - `participants`: array of strings
+  - `state`: string
+  - `hostname`: string
+  - `address`: string
+  - `login`: string
+  - `cluster`: string
+  - `kube_cluster`: string
+  - `host_user`: string
+  - `host_roles`: array of strings
+
+Configuration example that denies list/read access to all sessions you are a participant of:
+
+```yaml
+spec:
+  deny:
+    rules:
+    - resources: [session_tracker]
+      verbs: [list, read]
+      where: 'contains(tracker.participants, user.metadata.name)'
+```
+
+### Effects on the default `auditor` role
+
+To preserve earlier functionality from the legacy session metadata system, we should allow the auditor to list all active sessions by default but without the ability to join them. This involves adding this section to the predefined role definition:
```

Comment:
This looks good.

I'm wondering if we should make an exception to the "Teleport never modifies the preset roles if they already exist" rule and ensure we add this allow rule to the auditor preset.

If we don't do this, every customer who has upgraded from v9 will get the new behavior and their auditors will never be able to see all sessions, which is something that worked out of the box in v8 and earlier.

Reply:
Hmm, yeah, that said I'm a bit icky on automagically editing roles that the customer expects something to be doing, that feels like somewhat of a breaking change in a minor.

Reply:
This would feel more okay if the inbuilt roles were immutable and everyone created new roles but many customers just modify existing roles to fit their purpose and reuse them.

Reply:
Okay, here's our path forward Joel:

- On startup, if `auditor` doesn't exist, we create it just like we do today but with this new allow rule.
- If the `auditor` role does exist, but has not been modified by the user, we should add the new allow rule so that the role functions the way it did in Teleport 9 and earlier.

You'll have to determine how to check whether the role has been modified by the user in order for this logic to kick in.

The cloud team is starting an effort next month to make the preset roles _immutable to users_. When this work is completed, users will be unable to modify the presets, and teleport will reserve the right to modify them during upgrades, which means you will be able to remove the custom logic for detecting whether the user modified the role.

Reply:
Sounds good Zac. Added a couple of sentences to the RFD to detail this.